File: neetcode/problems/leetcode1700/solution.py
from typing import List


class Node:
    def __init__(self, val: int):
        self.value = val
        self.next = None


class LinkedQueue:
    def __init__(self, queue: List[int]):
        self.length = 1
        self.head = Node(queue[0])
        self.tail = self.head

        for i in range(1, len(queue)):
            self.enqueue(queue[i])

    def enqueue(self, val: int):
        self.tail.next = Node(val)
        self.tail = self.tail.next
        self.length += 1

    def dequeue(self):
        val = self.head.value
        self.head = self.head.next
        self.length -= 1
        return val

    def peek(self):
        return self.head.value


class Solution:
    def countStudents(self, students: List[int], sandwiches: List[int]) -> int:
        studentsQueue = LinkedQueue(students)
        sandwichesQueue = LinkedQueue(sandwiches)
        count = 0

        while count != sandwichesQueue.length:
            if studentsQueue.peek() == sandwichesQueue.peek():
                studentsQueue.dequeue()
                sandwichesQueue.dequeue()
                count = 0
            else:
                studentsQueue.enqueue(studentsQueue.dequeue())
                count += 1

        return count

File: neetcode/problems/leetcode1700/solution.js
class Node {
	constructor(val) {
		this.value = val;
		this.next = null;
	}
}

class QueueList {
	constructor(queue) {
		this.length = 1;
		this.head = new Node(queue[0]);
		this.tail = this.head;

		for (let i = 1; i < queue.length; i++) {
			this.enqueue(queue[i]);
		}
	}

	peek() {
		return this.head.value;
	}

	enqueue(value) {
		this.tail.next = new Node(value);
		this.tail = this.tail.next;
		this.length++;
	}

	dequeue() {
		const { value } = this.head;
		this.head = this.head.next;
		this.length--;

		return value;
	}
}

const countStudents = (students, sandwiches) => {
	const studentsQueue = new QueueList(students);
	const sandwitchesQueue = new QueueList(sandwiches);

	let count = 0;

	while (count !== sandwitchesQueue.length) {
		if (studentsQueue.peek() === sandwitchesQueue.peek()) {
			studentsQueue.dequeue();
			sandwitchesQueue.dequeue();
			count = 0;
		} else {
			studentsQueue.enqueue(studentsQueue.dequeue());
			count++;
		}
	}

	return count;
};
